### ISSUE TITLE
Regex for Mineplex

### ISSUE THREAD

**Issue by ReOp14** [ORelio/Minecraft-Console-Client#201]
Yeah, its me and my avatar of my cat again.
Ok so, I have always been fairly terrible at regex and the main server I really go on is Mineplex
I always use this [Regex Code](https://regex101.com/r/iC6mF0/1) for the main config file:
    `^.*(?:Owner|MPS Host|MPS Co-Host|LEGEND|HERO|ULTRA|TITAN|TRAINEE|BUILDER) ([a-zA-Z0-9_]+) (.+)$`

The problem with [this regex code](https://regex101.com/r/iC6mF0/1) is that it only captures people with **Ranks** or other special tags.
Ex: `TITAN ReOp Text, this is text.` and it has no problem with that or with any other rank in the [Regex Code](https://regex101.com/r/iC6mF0/1). But, players **without** ranks or special tags don't get detected by the [regex expression](https://regex101.com/r/iC6mF0/1).
Ex: `DeOp Hey guys im new 2 mineplex server` DeOp simply has no rank and therefor dosn't get affected by the Regex. (Which is the problem)

There is another problem though, if you're **Dead** in game, it will show up as:
`Dead LEGEND ORelio Regex is awesome!` or **without** a rank:`Dead Notch Minecraft is getting old`
(Thats why there is **.*** in the first part of the Regex)

In conclusion, the [Regex Expression](https://regex101.com/r/iC6mF0/1) `^.*(?:Owner|MPS Host|MPS Co-Host|LEGEND|HERO|ULTRA|TITAN|TRAINEE|BUILDER) ([a-zA-Z0-9_]+) (.+)$`
doesn't detect Players without a syntax before the name. So I need a fix for that if you know how.

Thanks!

**Comment by Pokechu22** [comment]
You can use `?` to make something optional, for instance like this:

     ^.*(?:Dead )?(?:Owner |MPS Host |MPS Co-Host |LEGEND |HERO |ULTRA |TITAN |TRAINEE |BUILDER )?([a-zA-Z0-9_]+) (.+)$ 

Note that I also added a space after each rank and removed the space between the ranks group and the name, since the space won't be there if there isn't a rank.

**Comment by ReOp14** [comment]
Thanks @Pokechu22 for the response! That Regex worked, all I had to do is remove the .* at the start of the regex.

**Comment by ReOp14** [comment]
Hey I found a further problem with the Regex @Pokechu22, 
`^(?:Dead )?(?:Owner |MPS Host |MPS Co-Host |LEGEND |HERO |ULTRA |TITAN |TRAINEE |BUILDER )?([a-zA-Z0-9_]+) (.+)$ `
If there is a simple message like: `announcement free ranks xD`
It will think it is a regex, https://regex101.com/r/kM7oU0/1

**Comment by Pokechu22** [comment]
> Although the regex might not be just "announcement" so don't try and match for that.

Unfortunately I can't think of any other way to detect those specific messages and exclude them since they're formatted the same - there may be one but I can't think of it right now.  So here's a regex that simply fails if the message starts with `Announcement`:

    ^(?!Announcement)(?:Dead )?(?:Owner |MPS Host |MPS Co-Host |LEGEND |HERO |ULTRA |TITAN |TRAINEE |BUILDER )?([a-zA-Z0-9_]+) (.+)$

It works with the negative lookahead, `?!`, meaning that if the message starts with `Announcement` it will not match.  You can add additional terms with `|` as needed (eg, `(?!Announcement|Alert)`).

**Comment by ReOp14** [comment]
Yeah that sounds like what I have to do, unfortunately, the Regex can't tell if its server text or a player chatting.

**Comment by Pokechu22** [comment]
Theoretically it might be possible to distinguish it using the [position field of the chat message packet](http://wiki.vg/Protocol#Chat_Message_.28clientbound.29), in that chat messages would have it set to chat and announcements would be system message, but most servers don't implement that field (since there's no good way of doing it through the bukkit API).  You could maybe get mineplex to switch to using the proper position and then detecting it that way, but I doubt it.

**Comment by ORelio** [comment]
Thanks @Pokechu22 for your help.
Yes, most messages are not properly tagged as "System" messages so that's not really useful.
You may hide system messages with `showsystemmessages=false` in bot.
Messages properly tagged as "System" will be ignored and not displayed nor processed.

**Comment by ReOp14** [comment]
Im at school right now so I can't try out the showsytemmessages. Well I so know mineplex has a few font changes but maybe if there was color detection in chat messages, we can detect if it was by a player or just chat text.

**Comment by ORelio** [comment]
Regex in config allows matching private/public messages.
You should put your message format detection regex in config instead of placing it in matches file.

**Comment by ReOp14** [comment]
It seems I was wrong about 
> Also, not sure if this is a bug, but if the regex in the Config file can match a regex in your matches file it will overwrite the regex and use the config's regex. If ORelio can fix this, that should be the easiest solution.

What I did mean by that is the INI file regex overrides the regex in the matches file, which is incorrect.
But since I know this now, I'll just manually convert all my "match=" to "regex=" which shouldn't be much of a problem, but still it will help a lot if there was a easier way to detect chat or player messages correctly.

**Comment by ReOp14** [comment]
Actually I think I am actually correct (even though I said I was incorrect) about my last comment because im trying to test right now and I cannot do anything while using the current regex. So here is the situation:
`^(?:Dead )?(?:Owner |MPS Host |MPS Co-Host |LEGEND |HERO |ULTRA |TITAN |TRAINEE |BUILDER )?([a-zA-Z0-9_]+) (.+)$`
 is my regex in my INI main file. And (a) regex in matches file:
```
regex=Welcome to Mineplex Private Servers!
actionother=script loginMPS
```
The regex in my matches file doesn't ever get detected whenever I switch the to the Regex in the INI file compared to my old Regex in the INI file. Also, it doesn't get detected if its match=Welcome to Mineplex Private Servers! Darn Regex!

**Comment by ORelio** [comment]
That's because the regex in config will match it as a public message.
So this is not a "other" message but a public message sent by "Welcome".
Also you don't need a regex here, "match=" instead of "regex=" is fine.

**Comment by ReOp14** [comment]
That's a good solution, but it is still a bug for a regex= to be overwritten by the regex in the main ini file?
And that may cause problems if someone does say `to Mineplex Private Servers`

**Comment by ORelio** [comment]
This is not a bug, this is the regex you are providing which causes server messages to be incorrectly categorized as public messages. You may improve it as @Pokechu22 hinted by blacklisting first words such as `Announcement` or `Welcome` so messages will be categorized as "other".

The "regex=" in matches file is for matching either message contents for public/private messages, or whole line for "other" messages. This is normal that the chat format regex is run first. Eg you want to match someone saying either "Hello" or "Bye" then, message shown on server is:
````
<Someone> Hello
````
First, the chat format regex will decompose it:
 - Category: public message
 - Username: `Someone`
 - Message: `Hello`

Then, and only then, it is matched by the regex in matches file:
 - Regex; `^Hello|Bye$`
 - Input: `Hello`
 - Result: Message is matching

Of course for "other" chat messages, the main chat format regex does not match anything so the full message is passed to the regex in matches file. This is the intended behavior.

**Comment by ReOp14** [comment]
Alright, so i'm at a standstill right now, and I think i'm going with the strategy of removing the first word. I really thank you for your help, ORelio and Pokechu22.

I do have another problem with removing the first word method. The msg sytax on Mineplex is:
 `Player1 > Player2 <message>`
And I have a Regex in my matches file: `regex=(ReOp|DeOp) > AntzBot script things`
And if I do remove the first word, it will be ReOp or DeOp and then the INI Regex will break this regex because I need it to detect who messaged AntzBot.

**Comment by ORelio** [comment]
You need to define a separate regex for private messages in main INI file.

**Comment by ReOp14** [comment]
What about a similar problem like the last comment.
regex=^1st Place - ([a-zA-Z0-9_]+)
If I remove 1st then the Regex dose not know if its 1st 2nd or 3rd.

**Comment by ORelio** [comment]
If PM syntax is `Player1 > Player2 <message>`
Then regex is `^([A-Za-z0-9_]+) > (?:[A-Za-z0-9_]+) (.+)$`
Live example at https://regex101.com/r/uM8hR2/1

**Comment by ReOp14** [comment]
I have been trying lots of things to get actionprivate= to work but I have not found anything.
I used your `^([A-Za-z0-9_]+) > (?:[A-Za-z0-9_]+) (.+)$` as the regex in the INI.
Here is a photo of the chat message: http://imgur.com/a/q2BBk

And this is my example I tested on: (I also tried "regex=")
`match=hello hello`
`actionprivate=send Hello!`

**Comment by ORelio** [comment]
Is the `private=` setting properly uncommented by removing the leading `#`?
If so, please enable the chat logger and make it log everything.
You should be able to see how messages are categorized and copy and paste a PM here.

**Comment by ReOp14** [comment]
Ok I'm trying to fix the private messaging. `2016-09-14 15:15:43 Chat ReOp: > AntzBot yo yo`
This is the text that was in the log file and I quickly realized that there was a colon. But I tried putting ":" in the regex and without and the match won't find anything.
`[Match]`
`match=yo yo`
`actionprivate=send @$1 says: $2`
Also, `private=^([A-Za-z0-9_]+): > (?:[A-Za-z0-9_]+) (.+)$` is in my INI file.

**Comment by Pokechu22** [comment]
Remember, the chatlog bot uses your regex to determine what type, displaying the type, followed by the sender, **followed by a colon**, followed by the message.  It attempts to parse it as a chat message first, and then as a private message.  Since it's saying it's a chat message, and not a PM, that means that it's seeing it as a chat message.

So, the issue here is that private messages are being matched by the chat regex.

You could exclude things that look like PMs to your bot in the chat regex by putting `(?!> Botname)` after the username and before the message, making a regex like this:

    ^(?:Dead )?(?:Owner |MPS Host |MPS Co-Host |LEGEND |HERO |ULTRA |TITAN |TRAINEE |BUILDER )?([a-zA-Z0-9_]+) (?!> AntzBot)(.+)$

And then hopefully the message regex should be used and work.

**Comment by ReOp14** [comment]
I tried that Regex @Pokechu22, It still hasn't worked. I also tried removing the plubic Regex entirely and the same result happened. So its most likely not a issue with the plubic Regex.

`[Match]`
`match=hello hello` (also tried regex=)
`action=send hey`
`actionprivate=send actionprivate`
I try messaging it "hello hello" and "hello" and no result in console nor chat.
`private=^([A-Za-z0-9_]+) > (?:[A-Za-z0-9_]+) (.+)$`

**Comment by ReOp14** [comment]
This issue is getting boring and has gone long enough I think. Regex is pretty hard to understand. I can just test for a private message using regex= in the matches file. Pretty simple. Still learned alot! Thank you!

I might start another issue tomorrow if I don't understand how using randomness (like you used in the [Hangman game](https://github.com/ORelio/Minecraft-Console-Client/blob/master/MinecraftClient/ChatBots/HangmanGame.cs)) works.

**Comment by ORelio** [comment]
Well, good luck, have fun! ^^
